client/src/controllers/TagController.ts

```diff
@@ -18,24 +19,88 @@ declare global {
  */
 export class TagController extends Controller {
   static values = {
+    delay: { type: Number, default: 300 },
     options: { default: {}, type: Object },
-    url: String,
+    url: { default: '', type: String },
   };
 
   declare optionsValue: any;
-  declare urlValue: any;
+  declare urlValue: string;
+  declare delayValue: number;
+
+  private autocompleteAbort: AbortController | null = null;
+  private autocompleteLazy;
+
   tagit?: JQuery<HTMLElement>;
 
+  initialize() {
+    this.autocompleteLazy = debounce(
+      this.autocomplete.bind(this),
+      this.delayValue,
+    );
+  }
+
   connect() {
     const preprocessTag = this.cleanTag.bind(this);
 
+    const autocomplete = {
+      source: (request, response) => {
+        this.autocompleteLazy.cancel();
+        this.autocompleteLazy(request).then(response);
+      },
+    };
+
     $(this.element).tagit({
-      autocomplete: { source: this.urlValue },
+      autocomplete,
       preprocessTag,
       ...this.optionsValue,
     });
   }
 
+  disconnect() {
+    if (this.autocompleteAbort) {
+      this.autocompleteAbort.abort();
+      this.autocompleteAbort = null;
+    }
+  }
+
+  async autocomplete({ term }: { term: string }) {
+    if (this.autocompleteAbort) {
+      this.autocompleteAbort.abort();
+    }
+
+    this.autocompleteAbort = new AbortController();
+    const { signal } = this.autocompleteAbort;
+
+    try {
+      const url = new URL(this.urlValue, window.location.origin);
+      url.searchParams.set('term', term);
+
+      const fetchResponse = await fetch(url.toString(), {
+        headers: { Accept: 'application/json' },
+        method: 'GET',
+        signal,
+      });
+
+      const data = await fetchResponse.json();
+      return Array.isArray(data) ? data : [];
+    } catch (error) {
+      const errorMessage =
+        error instanceof DOMException && error.name === 'AbortError'
```

Comment:
No need to log the AbortError, this is not really an error as it's intentional, let's just return when this happens.